### ISSUE TITLE
Clarification about Android 6.0 MM permissions

### ISSUE THREAD

**Issue by mirko77** [dpa99c/cordova-diagnostic-plugin#96]
Hi there, this is not a bug but more of a clarification about an issue I am facing on Android 6.0 MM

I am building an app where I use the camera for photo, shooting videos and barcode scanning. I installed the appropriate plugins, and NONE of them added any camera permission on the Android manifest.

I read somewhere the permission is not needed any more because the camera now use intents. Fair enough, but if I DO NOT manually add these two lines in my manifest

```
 <uses-permission android:name="android.permission.CAMERA" />
 <uses-permission android:name="android.permission.FLASHLIGHT" />
```

I get permission denied and the camera permission does not even appear in the app settings. A call to `window.cordova.plugins.diagnostic.isCameraAuthorized()` will always return `DENIED_ALWAYS`, same for `cordova.plugins.diagnostic.requestRuntimePermission()` asking for the camera passing `cordova.plugins.diagnostic.runtimePermission.CAMERA`.

The weird thing is everything was working fine without the camera permission, then I did a fresh re-install and it all went belly up. I am not sure if this is due to some cordova-android (using 6.1.1) or plugin updates or what. I also tried a fork of the camera plugin, as others found the camera permission popup does not get triggered using the official one (https://github.com/apache/cordova-plugin-camera/commit/c12206ebc8ec337613e0751970536d692de6f572) but nothing changes...

Any thoughts?

btw, I am adding those permissions in using your own plugin `cordova-custom-config`

**Comment by dpa99c** [comment]
Any run-time permissions that you want to request **must** also be defined in the manifest via `<uses-permission/>` element, otherwise `requestRuntimePermission()` will always return `DENIED_ALWAYS`.  This is stated explicitly in the [Android permissions documentation](https://developer.android.com/training/permissions/requesting.html):

> On all versions of Android, your app needs to declare both the normal and the dangerous permissions it needs in its app manifest, as described in Declaring Permissions.

Your observed changed in behaviour may be explained by your upgrade to Cordova CLI 6.1.1 - depending what version you upgraded from. Cordova CLI 6.1.1 defaults to `cordova-android@5.2.0` (I think) which is based on API 23 aka Android 6.0 so adds run-time permission support to Cordova. If you had recent versions the official plugins installed, these contain a backward-compatibility shim to make them work with earlier cordova-android versions, but as soon as you have cordova-android@5+ installed, the run-time permissions code will kick in.

You can see [in the source code](https://github.com/apache/cordova-plugin-camera/blob/master/src/android/CameraLauncher.java#L112) that cordova-plugin-camera requires `CAMERA` and `READ_EXTERNAL_STORAGE` permissions. It should take care of requesting these permissions itself, however you can use this plugin to manually manage the requesting of these relevant permissions.

**Comment by mirko77** [comment]
Thanks for the info.

Cordova 6.1.1 adds Android platform 5.1.1

I can confirm the camera permission is NOT added by the camera plugin. Easy to check with:

`cordova create hello com.example.hello HelloWorld`
`cd hello`
`cordova platform add android`
`cordova plugin add cordova-plugin-camera`

the i run a `cordova prepare` or a `cordova run android --device` and the manifest is:

```
<?xml version='1.0' encoding='utf-8'?>
<manifest android:hardwareAccelerated="true" android:versionCode="1" android:versionName="0.0.1" package="com.example.hello" xmlns:android="http://schemas.android.com/apk/res/android">
    <supports-screens android:anyDensity="true" android:largeScreens="true" android:normalScreens="true" android:resizeable="true" android:smallScreens="true" android:xlargeScreens="true" />
    <uses-permission android:name="android.permission.INTERNET" />
    <application android:hardwareAccelerated="true" android:icon="@drawable/icon" android:label="@string/app_name" android:supportsRtl="true">
        <activity android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/activity_name" android:launchMode="singleTop" android:name="MainActivity" android:theme="@android:style/Theme.DeviceDefault.NoActionBar" android:windowSoftInputMode="adjustResize">
            <intent-filter android:label="@string/launcher_name">
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>
    <uses-sdk android:minSdkVersion="14" android:targetSdkVersion="23" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
</manifest>
```

**Comment by dpa99c** [comment]
Yep, it looks like while they've added code to [request CAMERA and READ_EXTERNAL_STORAGE permissions at runtime](https://github.com/apache/cordova-plugin-camera/blob/master/src/android/CameraLauncher.java#L112) to cordova-plugin-camera, they only have [WRITE_EXTERNAL_STORAGE permission in the plugin.xml](https://github.com/apache/cordova-plugin-camera/blob/master/plugin.xml#L70), which is why you're not seeing CAMERA in the resulting AndroidManifest.xml

So you can either continue to use `cordova-custom-config` to add the manifest entry for CAMERA, or fork cordova-plugin-camera and add it to their plugin.xml - you might want to also create a pull request to get them to merge this back into the main plugin repo.
